### PR TITLE
Fixes, refactoring, and import sorting

### PR DESCRIPTION
**Pull request by anonymoose2**

### CODE REVIEW

File: intranet/apps/eighth/models.py [hunk 1]
```diff
@@ -570,78 +630,131 @@ def previous_blocks(self, quantity=-1):
...
+        Return:
+            The width in pixels of the block letter.
+
+        """
         return (len(self.block_letter) - 1) * 6 + 15
 
     @property
-    def short_text(self):
+    def short_text(self) -> str:
         """Display the date and block letter (mm/dd B, for example: '9/1 B')
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns the date and block letter for this block.
        
         It is returned in the format of  MM/DD B, like "9/1 B"
```

File: intranet/apps/eighth/models.py [hunk 2]
```diff
@@ -570,78 +630,131 @@ def previous_blocks(self, quantity=-1):
...
+            The date and block letter.
+
         """
         return "{} {}".format(self.date.strftime("%m/%d"), self.block_letter)
 
     @property
-    def is_this_year(self):
-        """Return whether the block occurs after September 1st of this school year."""
+    def is_this_year(self) -> bool:
+        """Return whether the block occurs after September 1st of this school year.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Return whether the block occurs during this school year.
```

File: intranet/apps/eighth/models.py [hunk 3]
```diff
@@ -570,78 +630,131 @@ def previous_blocks(self, quantity=-1):
...
+
+        Returns:
+            Whteher the block occurs during this school year.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
            Whether the block occurs during this school year.
```

File: intranet/apps/eighth/models.py [hunk 4]
```diff
@@ -570,78 +630,131 @@ def previous_blocks(self, quantity=-1):
...
+
+        """
         return is_current_year(datetime.datetime.combine(self.date, datetime.time()))
 
     @property
-    def formatted_date(self):
+    def formatted_date(self) -> str:
         """Returns the block date, formatted according to settings.EIGHTH_BLOCK_DATE_FORMAT.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns the block date, formatted according to ``settings.EIGHTH_BLOCK_DATE_FORMAT``.
```

File: intranet/apps/eighth/models.py [hunk 5]
```diff
@@ -657,7 +770,9 @@ class Meta:
 class EighthScheduledActivityManager(Manager):
     """Model Manager for EighthScheduledActivity."""
 
-    def for_sponsor(self, sponsor, include_cancelled=False):
+    def for_sponsor(self, sponsor: EighthSponsor,
+                    include_cancelled: bool = False) -> Union[QuerySet,
+                                                              Collection["EighthScheduledActivity"]]:  # pylint: disable=unsubscriptable-object
         """Return a QueryList of EighthScheduledActivities where the given EighthSponsor is
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns a QuerySet of EighthScheduledActivities where the given EighthSponsor is
```

File: intranet/apps/eighth/models.py [hunk 6]
```diff
@@ -657,7 +770,9 @@ class Meta:
 class EighthScheduledActivityManager(Manager):
     """Model Manager for EighthScheduledActivity."""
 
-    def for_sponsor(self, sponsor, include_cancelled=False):
+    def for_sponsor(self, sponsor: EighthSponsor,
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]: This method is missing an `Args` block.

File: intranet/apps/eighth/models.py [hunk 7]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
         return name_with_flags
 
-    def get_true_sponsors(self):
+    def get_true_sponsors(self) -> Union[QuerySet, Collection[EighthSponsor]]:  # pylint: disable=unsubscriptable-object
         """Get the sponsors for the scheduled activity, taking into account activity defaults and
-        overrides."""
+        overrides.
+
+        Returns:
+            The sponsors for the scheduled activity
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
            The sponsors for this scheduled activity.
```

File: intranet/apps/eighth/models.py [hunk 8]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
+        """
         cancelled_str = " (Cancelled)" if self.cancelled else ""
         name_with_flags = self.activity._name_with_flags(True, self.title) + cancelled_str  # pylint: disable=protected-access
         if self.special and not self.activity.special:
             name_with_flags = "Special: " + name_with_flags
         return name_with_flags
 
-    def get_true_sponsors(self):
+    def get_true_sponsors(self) -> Union[QuerySet, Collection[EighthSponsor]]:  # pylint: disable=unsubscriptable-object
         """Get the sponsors for the scheduled activity, taking into account activity defaults and
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Retrieves the sponsors for the scheduled activity, taking into account activity defaults and
```

File: intranet/apps/eighth/models.py [hunk 9]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
+
+        Returns:
+            The sponsors for the scheduled activity
+
+        """
         return self.sponsors.all() or self.activity.sponsors.all()
 
-    def user_is_sponsor(self, user):
+    def user_is_sponsor(self, user: "get_user_model()") -> bool:
         """Return whether the given user is a sponsor of the activity.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether the given user is a sponsor of the activity.
```

File: intranet/apps/eighth/models.py [hunk 10]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
 
         Returns:
-            Boolean
+            Whether the user is a sponsor of the activity.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
            Whether the given user is a sponsor of the activity.
```

File: intranet/apps/eighth/models.py [hunk 11]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
+
+        Returns:
+            The sponsors for the scheduled activity
+
+        """
         return self.sponsors.all() or self.activity.sponsors.all()
 
-    def user_is_sponsor(self, user):
+    def user_is_sponsor(self, user: "get_user_model()") -> bool:
         """Return whether the given user is a sponsor of the activity.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]: Missing an `Args` block.

File: intranet/apps/eighth/models.py [hunk 12]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
         Returns:
-            Boolean
+            Whether the user is a sponsor of the activity.
+
         """
         return self.get_true_sponsors().filter(user=user).exists()
 
-    def get_true_rooms(self):
+    def get_true_rooms(self) -> Union[QuerySet, Collection[EighthRoom]]:  # pylint: disable=unsubscriptable-object
         """Get the rooms for the scheduled activity, taking into account activity defaults and
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Retrieves the rooms for the scheduled activity, taking into account
        activity defaults and overrides.
```

File: intranet/apps/eighth/models.py [hunk 13]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
-        overrides."""
+        overrides.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
```

File: intranet/apps/eighth/models.py [hunk 14]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
+
+        Returns:
+            The true room list of the scheduled activity.
+
+        """
         return self.rooms.all() or self.activity.rooms.all()
 
-    def get_true_capacity(self):
+    def get_true_capacity(self) -> int:
         """Get the capacity for the scheduled activity, taking into account activity defaults and
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Retrieves the capacity for the scheduled activity, taking into
        account activity defaults and overrides.
```

File: intranet/apps/eighth/models.py [hunk 15]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
-        overrides."""
+        overrides.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
```

File: intranet/apps/eighth/models.py [hunk 16]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
-            bool: Whether this activity runs both blocks.
+            Whether this scheduled activity runs both blocks.
 
         """
         return self.both_blocks or self.activity.both_blocks
 
-    def get_restricted(self):
-        """Get whether this scheduled activity is restricted."""
+    def get_restricted(self) -> bool:
+        """Get whether this scheduled activity is restricted.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Gets whether this scheduled activity is restricted.
```

File: intranet/apps/eighth/models.py [hunk 17]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+        Returns:
+            Whether this scheduled activity is restricted.
+
+        """
         return self.restricted or self.activity.restricted
 
-    def get_sticky(self):
-        """Get whether this scheduled activity is sticky."""
+    def get_sticky(self) -> bool:
+        """Get whether this scheduled activity is sticky.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Gets whether this scheduled activity is sticky.
```

File: intranet/apps/eighth/models.py [hunk 18]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+        Returns:
+            Whether this scheduled activity is sticky.
+
+        """
         return self.sticky or self.activity.sticky
 
-    def get_finance(self):
-        """Get whether this activity has an account with the finance office."""
+    def get_finance(self) -> str:
+        """Get the name of this activity's account with the finance office.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Retrieves the name of this activity's account with the
        finance office, if any.
```

File: intranet/apps/eighth/models.py [hunk 19]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
         cancelled_str = " (Cancelled)" if self.cancelled else ""
         act_name = self.activity.name + cancelled_str
         if self.special and not self.activity.special:
             act_name = "Special: " + act_name
         return act_name if not self.title else "{} - {}".format(act_name, self.title)
 
     @property
-    def title_with_flags(self):
+    def title_with_flags(self) -> str:
         """Gets the title for the activity, appending the title of the scheduled activity to the
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Gets the title for the activity, appending the title of the
        scheduled activity to the activity's name and flags.
```

File: intranet/apps/eighth/models.py [hunk 20]
```diff
@@ -743,56 +858,82 @@ class EighthScheduledActivity(AbstractBaseEighthModel):
...
-        activity's name and flags."""
+        activity's name and flags.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
```

File: intranet/apps/eighth/models.py [hunk 21]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+        Returns:
+            The name of this activity's account with the finance office.
+
+        """
         return self.activity.finance
 
-    def get_administrative(self):
-        """Get whether this scheduled activity is administrative."""
+    def get_administrative(self) -> bool:
+        """Get whether this scheduled activity is administrative.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether this scheduled activity is administrative.
```

File: intranet/apps/eighth/models.py [hunk 22]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+
+        Returns:
+            Whether this activity is administrative
+
+        """
         return self.administrative or self.activity.administrative
 
-    def get_special(self):
+    def get_special(self) -> bool:
         """Get whether this scheduled activity is special, checking the
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether this scheduled activity is special.
```

File: intranet/apps/eighth/models.py [hunk 23]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
-           activity-level special settings.
+        activity-level special settings.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
```

File: intranet/apps/eighth/models.py [hunk 24]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+
+        Returns: Whether this scheduled activity is special.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        Returns:
            Whether this scheduled activity is special.
```

File: intranet/apps/eighth/models.py [hunk 25]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+
         """
         return self.special or self.activity.special
 
-    def is_full(self):
-        """Return whether the activity is full."""
+    def is_full(self) -> bool:
+        """Return whether the scheduled activity is full.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether the scheduled activity is full.
```

File: intranet/apps/eighth/models.py [hunk 26]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+            Whether this scheduled activity is full.
+
+        """
         capacity = self.get_true_capacity()
         return capacity != -1 and self.eighthsignup_set.count() >= capacity
 
-    def is_almost_full(self):
-        """Return whether the activity is almost full (>90%)."""
+    def is_almost_full(self) -> bool:
+        """Return whether the scheduled activity is almost full (>=90%).
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether the scheduled activity is almost full (>=90%).
```

File: intranet/apps/eighth/models.py [hunk 27]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+            Whether this scheduled activity is at least 90% full.
+
+        """
         capacity = self.get_true_capacity()
         return capacity != -1 and self.eighthsignup_set.count() >= (0.9 * capacity)
 
-    def is_overbooked(self):
-        """Return whether the activity is overbooked."""
+    def is_overbooked(self) -> bool:
+        """Return whether the activity is overbooked (>100%) capacity.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether the activity is overbooked (>100%) capacity.
```

File: intranet/apps/eighth/models.py [hunk 28]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+        Returns:
+            Whether this scheduled activity is overbooked.
+
+        """
         capacity = self.get_true_capacity()
         return capacity != -1 and self.eighthsignup_set.count() > capacity
 
-    def is_too_early_to_signup(self, now=None):
+    def is_too_early_to_signup(self, now: datetime.datetime = None) -> bool:
         """Return whether it is too early to sign up for the activity if it is a presign (48 hour
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether it is too early to sign up for the activity
        if it is a presign.
        
        This contains the 48 hour presign logic.
```

File: intranet/apps/eighth/models.py [hunk 29]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
-        logic is here)."""
+        logic is here).
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
```

File: intranet/apps/eighth/models.py [hunk 30]
```diff
@@ -866,14 +1050,21 @@ def is_too_early_to_signup(self, now=None):
 
         return now < (activity_date - presign_period)
 
-    def has_open_passes(self):
-        """Return whether there are passes that have not been acknowledged."""
+    def has_open_passes(self) -> bool:
+        """Return whether there are passes that have not been acknowledged.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether there are passes that have not been acknowledged.
```

File: intranet/apps/eighth/models.py [hunk 31]
```diff
@@ -866,14 +1050,21 @@ def is_too_early_to_signup(self, now=None):
...
+        Returns:
+            Whether this activity has open passes.
+
+        """
         return self.eighthsignup_set.filter(after_deadline=True, pass_accepted=False).exists()
 
-    def _get_viewable_members(self, user):
+    def _get_viewable_members(self, user: "get_user_model()") -> Union[QuerySet,
+                                                                       Collection["get_user_model()"]]:  # pylint: disable=unsubscriptable-object
         """Get a (unsorted) QuerySet of the members that you have permission to view.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Get an unsorted QuerySet of the members that you have permission to view.
```

File: intranet/apps/eighth/models.py [hunk 32]
```diff
@@ -884,27 +1075,34 @@ def _get_viewable_members(self, user):
                 q |= Q(id=user.id)
             return self.members.filter(q)
 
-    def get_viewable_members(self, user=None):
+    def get_viewable_members(self,
+                             user: "get_user_model()" = None) -> Union[QuerySet,
+                                                                       Collection["get_user_model()"]]:  # pylint: disable=unsubscriptable-object
         """Get a QuerySet of the members that you have permission to view, sorted alphabetically.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns a QuerySet of the members that you have permission to view, sorted alphabetically.
```

File: intranet/apps/eighth/models.py [hunk 33]
```diff
@@ -884,27 +1075,34 @@ def _get_viewable_members(self, user):
...
-        Returns: QuerySet of the members that you have permission to view, sorted alphabetically.
+        Returns:
+            QuerySet of the members that you have permission to view, sorted alphabetically.
 
         """
         return self._get_viewable_members(user).order_by("last_name", "first_name")
 
-    def get_viewable_members_serializer(self, request):
+    def get_viewable_members_serializer(self, request) -> Union[QuerySet, Collection["get_user_model()"]]:  # pylint: disable=unsubscriptable-object
         """Given a request, get an unsorted QuerySet of the members that the requesting user
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Given a request, returns an unsorted QuerySet of the members that the requesting user
```

File: intranet/apps/eighth/models.py [hunk 34]
```diff
@@ -884,27 +1075,34 @@ def _get_viewable_members(self, user):
                 q |= Q(id=user.id)
             return self.members.filter(q)
 
-    def get_viewable_members(self, user=None):
+    def get_viewable_members(self,
+                             user: "get_user_model()" = None) -> Union[QuerySet,
+                                                                       Collection["get_user_model()"]]:  # pylint: disable=unsubscriptable-object
         """Get a QuerySet of the members that you have permission to view, sorted alphabetically.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]: Missing `Args` block.

File: intranet/apps/eighth/models.py [hunk 35]
```diff
@@ -866,14 +1050,21 @@ def is_too_early_to_signup(self, now=None):
...
+        Returns:
+            Whether this activity has open passes.
+
+        """
         return self.eighthsignup_set.filter(after_deadline=True, pass_accepted=False).exists()
 
-    def _get_viewable_members(self, user):
+    def _get_viewable_members(self, user: "get_user_model()") -> Union[QuerySet,
+                                                                       Collection["get_user_model()"]]:  # pylint: disable=unsubscriptable-object
         """Get a (unsorted) QuerySet of the members that you have permission to view.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]: Missing `Args` block.

File: intranet/apps/eighth/models.py [hunk 36]
```diff
@@ -806,55 +947,98 @@ def get_true_capacity(self):
...
+        """
         capacity = self.get_true_capacity()
         return capacity != -1 and self.eighthsignup_set.count() > capacity
 
-    def is_too_early_to_signup(self, now=None):
+    def is_too_early_to_signup(self, now: datetime.datetime = None) -> bool:
         """Return whether it is too early to sign up for the activity if it is a presign (48 hour
-        logic is here)."""
+        logic is here).
+
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]: Missing `Args` block

File: intranet/apps/eighth/models.py [hunk 37]
```diff
@@ -884,27 +1075,34 @@ def _get_viewable_members(self, user):
...
+            Unsorted QuerySet of the members that you have permssion to view.
 
         """
         return self._get_viewable_members(request.user)
 
-    def get_hidden_members(self, user=None):
+    def get_hidden_members(self,
+                           user: "get_user_model()" = None) -> Union[QuerySet,
+                                                                     Collection["get_user_model()"]]:  # pylint: disable=unsubscriptable-object
         """Get a QuerySet of the members that you do not have permission to view.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]: Missing `Args` block

File: intranet/apps/eighth/models.py [hunk 38]
```diff
@@ -884,27 +1075,34 @@ def _get_viewable_members(self, user):
...
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns a QuerySet of the members that you do not have permission to view.
```

File: intranet/apps/eighth/models.py [hunk 39]
```diff
@@ -916,16 +1114,16 @@ def get_hidden_members(self, user=None):
...
+    def get_both_blocks_sibling(self) -> Optional["EighthScheduledActivity"]:
         """If this is a both-blocks activity, get the other EighthScheduledActivity
-           object that occurs on the other block.
+        object that occurs on the other block.
+
+        both_blocks means A and B block, NOT all of the blocks on that day.
 
-           both_blocks means A and B block, NOT all of the blocks on that day.
+        Returns:
+            EighthScheduledActivity object if found
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
            EighthScheduledActivity object if found.
```

File: intranet/apps/eighth/models.py [hunk 40]
```diff
@@ -916,16 +1114,16 @@ def get_hidden_members(self, user=None):
...
+            None if the activity cannot have a sibling
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
            ``None``, if the activity cannot have a sibling.
```

File: intranet/apps/eighth/models.py [hunk 41]
```diff
@@ -943,24 +1141,34 @@ def get_both_blocks_sibling(self):
...
         """
         for waitlist in waitlists:
             email_send_task.delay("eighth/emails/waitlist.txt", "eighth/emails/waitlist.html", {"activity": waitlist.scheduled_activity},
                                   "Open Spot Notification", [waitlist.user.primary_email_address])
 
     @transaction.atomic
-    def add_user(self, user, request=None, force=False, no_after_deadline=False, add_to_waitlist=False):
+    def add_user(self, user: "get_user_model()", request: Optional[HttpRequest] = None, force: bool = False, no_after_deadline: bool = False,
+                 add_to_waitlist: bool = False):
         """Sign up a user to this scheduled activity if possible. This is where the magic happens.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Signs up a user to this scheduled activity if possible. This is where the magic happens.
```

File: intranet/apps/eighth/models.py [hunk 42]
```diff
@@ -943,24 +1141,34 @@ def get_both_blocks_sibling(self):
...
 
         Raises an exception if there's a problem signing the user up
         unless the signup is forced.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        unless the signup is forced and the requesting user has permission.
```

File: intranet/apps/eighth/models.py [hunk 43]
```diff
@@ -1278,12 +1486,12 @@ def __str__(self):
 class EighthSignupManager(Manager):
     """Model manager for EighthSignup."""
 
-    def create_signup(self, user, scheduled_activity, **kwargs):
+    def create_signup(self, user: "get_user_model()", scheduled_activity: "EighthScheduledActivity", **kwargs) -> "EighthSignup":
         """Creates an EighthSignup for the given user in the given activity (after checking for duplicate signups).
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Creates an EighthSignup for the given user in the given activity after checking for duplicate signups.
        
        This raises an error if there are duplicate signups.
```

File: intranet/apps/eighth/models.py [hunk 44]
```diff
@@ -1373,17 +1581,27 @@ def validate_unique(self, *args, **kwargs):  # pylint: disable=arguments-differ
...
-            bool: Whether there is another EighthSignup for the same user in the same block.
+            Whether there is another EighthSignup for the same user in the same block.
 
         """
         return EighthSignup.objects.exclude(pk=self.pk).filter(user=self.user, scheduled_activity__block=self.scheduled_activity.block).exists()
 
-    def remove_signup(self, user=None, force=False, dont_run_waitlist=False):
-        """Attempt to remove the EighthSignup if the user has permission to do so."""
+    def remove_signup(self, user: "get_user_model()" = None, force: bool = False, dont_run_waitlist: bool = False) -> str:
+        """Attempt to remove the EighthSignup if the user has permission to do so.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Attempts to remove the EighthSignup if the user has permission to do so.
```

File: intranet/apps/eighth/models.py [hunk 45]
```diff
@@ -1430,8 +1648,13 @@ def reject_pass(self):
         self.pass_accepted = True
         self.save(update_fields=["was_absent", "pass_accepted"])
 
-    def in_clear_absence_period(self):
-        """Is the block for this signup in the clear absence period?"""
+    def in_clear_absence_period(self) -> bool:
+        """Is the block for this signup in the clear absence period?
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether the block for this signup is in the clear absence period.
```

File: intranet/apps/eighth/models.py [hunk 46]
```diff
@@ -1451,43 +1674,44 @@ class Meta:
 class EighthWaitlistManager(Manager):
     """Model manager for EighthWaitlist."""
 
-    def get_next_waitlist(self, activity):
+    def get_next_waitlist(self, activity: EighthScheduledActivity) -> Union[QuerySet,
+                                                                            Collection["EighthWaitlist"]]:  # pylint: disable=unsubscriptable-object
         """Return a QuerySet of all the EighthWaitlist objects for the given activity, ordered
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns a QuerySet of all the EighthWaitlist objects for the given
        activity, ordered by signup time.
```

File: intranet/apps/eighth/models.py [hunk 47]
```diff
@@ -1451,43 +1674,44 @@ class Meta:
...
         by signup time.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
```

File: intranet/apps/eighth/models.py [hunk 48]
```diff
@@ -1451,43 +1674,44 @@ class Meta:
...
-            QuerySet[EighthWaitlist]: A QuerySet of all the EighthWaitlist objects for the given activity,
+            A QuerySet of all the EighthWaitlist objects for the given activity,
             ordered by signup time.
 
         """
         return self.filter(scheduled_activity_id=activity.id).order_by("time")
 
-    def check_for_prescence(self, activity, user):
+    def check_for_prescence(self, activity: EighthScheduledActivity, user: "get_user_model()") -> bool:
         """Returns True if the given user is in a waitlist for the given activity.
```
Comment (theo-o) [inline intranet/apps/eighth/models.py]:
```suggestion
        """Returns whether the given user is in a waitlist for the given activity.
```

File: intranet/apps/eighth/models.py [hunk 49]
```diff
@@ -132,15 +134,14 @@ class EighthRoom(AbstractBaseEighthModel):
     unique_together = (("name", "capacity"),)
 
     @staticmethod
-    def total_capacity_of_rooms(rooms):
+    def total_capacity_of_rooms(rooms: Iterable["EighthRoom"]) -> int:
```
[the remark this answers is not in the document]
Comment (anonymoose2) [reply intranet/apps/eighth/models.py]: It's a [forward reference](https://www.python.org/dev/peps/pep-0484/#forward-references). `EighthRoom` has not been defined yet, so I can't reference it.

**Comment by anonymoose2** [comment]
> _I still do not like `fetch_activity_list_with_metadata`_

Me neither. There's a reason that #454 is *not* closed by this PR.

**Pull request merged**